### ISSUE TITLE
Ansible Limit not working properly

### ISSUE THREAD

**Issue by milad-soufastai** [ansible/ansible#14587]
##### Ansible Version:

ansible 2.0.0.2
  config file =
  configured module search path = Default w/o overrides
##### Ansible Configuration:

Please mention any settings you've changed/added/removed in ansible.cfg
(or using the ANSIBLE_\* environment variables).
##### Environment:

Mac OS X 10.11.3
##### Summary:

The limit flag seems to be broken in its current implementation
##### Steps To Reproduce:

```
ansible-playbook provisioning.yml --list-hosts
returns

  play #1 (lincs-ald):  TAGS: [ald_copy]
    pattern: [u'lincs-ald']
    hosts (1):
      aldprod

  play #2 (ALD):    TAGS: [ald]
    pattern: [u'ALD']
    hosts (1):
      ald_local

(Trimmed for length)

ansible-playbook provisioning.yml -l=aldprod --list-hosts
```
##### Expected Results:

```
playbook: provisioning.yml

  play #1 (lincs-ald):  TAGS: [ald_copy]
    pattern: [u'lincs-ald']
    hosts (1):
      aldprod

  play #2 (ALD):    TAGS: [ald]
    pattern: [u'ALD']
    hosts (0):
```
##### Actual Results:

ERROR! Specified --limit does not match any hosts 

```
ansible-playbook provisioning.yml -l=aldprod --list-hosts
ERROR! Specified --limit does not match any hosts
```
##### Workaround:

My temporary workaround is to run `ansible-playbook provisioning.yml -l=:aldprod --list-hosts`
The leading : seems to get the limit flag limiting to that single host. I've also tested this using more then 1 host and the same issue exists.

**Comment by sivel** [comment]
What does your `hosts` look like in your playbook?  Would be beneficial to have your playbook to look at.

**Comment by milad-soufastai** [comment]
```
- hosts: lincs-ald
  remote_user: "{{ ansible_user }}"
  roles:
  - ald_sync
  tags:
    - ald_copy

- hosts: ALD
  remote_user: root
  roles:
  - ald-init
  tags:
    - ald
``` -- Here's that part of the Playbook. It is simply running 2 different roles, one against a remote server, and one against a vagrant box (Takes a backup of the remote, restores to vagrant)

**Comment by milad-soufastai** [comment]
Here is a more complete example:

Inventory:

```
[hubot]
miladbox ansible_host:192.168.60.1
```

Playbook:

```
- hosts: hubot
  become_user: root
  become: yes
  tasks:
    - name: Update Yum
      yum: name=* state=latest
```

Command: `ansible-playbook example.yml --list-hosts`

```
playbook: example.yml

  play #1 (hubot):  TAGS: []
    pattern: [u'hubot']
    hosts (1):
      miladbox
```

Command after limit: `ansible-playbook example.yml -l=miladbox --list-hosts`

```
ERROR! Specified --limit does not match any hosts
```

with workaround posted above: `ansible-playbook example.yml -l=:miladbox --list-hosts`

```
playbook: example.yml

  play #1 (hubot):  TAGS: []
    pattern: [u'hubot']
    hosts (1):
      miladbox
```

**Comment by sivel** [comment]
So the problem is in how the option parser works and the difference between a short opt and a long opt.

The 2 options for limit are `-l` and `--limit`.  When using a short opt such as `-l` you do not use an equal sign, when using a long opt such as `--limit` you can either use or not use the `=`.

What happens when you use `-l` with an equal, is that the value becomes `=miladbox` where you were expecting just `miladbox`.  This is why using `-l=:miladbox`works, because it creates a hostlist that contains 2 hosts, one being `=` and the other `miladbox`.

In short, when using `-l` you should specify `-l miladbox` but when using `--limit` you can use `--limit=miladbox` or `--limit miladbox`

This is not something within our control, as this is how the optparse library in python operates.

**Comment by milad-soufastai** [comment]
Thanks for the explanation. Makes sense!
